tests/bluetooth/iso/modules/iso_broadcast_src.c

```diff
@@ -0,0 +1,400 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/iso.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/drivers/gpio.h>
+#include <stdlib.h>
+#include <ctype.h>
+#include <getopt.h>
+#include <unistd.h>
+
+#include "iso_broadcast_src.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(broadcast_src, CONFIG_ISO_TEST_LOG_LEVEL);
+
+K_THREAD_STACK_DEFINE(broadcaster_thread_stack, 4096);
+static struct k_thread broadcaster_thread;
+
+static const struct gpio_dt_spec led = GPIO_DT_SPEC_GET(DT_ALIAS(led1), gpios);
+
+#define BUF_ALLOC_TIMEOUT (10) /* milliseconds */
+#define BIG_SDU_INTERVAL_US (10000)
+
+NET_BUF_POOL_FIXED_DEFINE(bis_tx_pool, CONFIG_BIS_ISO_CHAN_COUNT_MAX,
+			  BT_ISO_SDU_BUF_SIZE(CONFIG_BT_ISO_TX_MTU), 8, NULL);
+
+static K_SEM_DEFINE(sem_big_cmplt, 0, CONFIG_BIS_ISO_CHAN_COUNT_MAX);
+static K_SEM_DEFINE(sem_big_term, 0, CONFIG_BIS_ISO_CHAN_COUNT_MAX);
+
+static uint16_t seq_num;
+static bool running;
+
+static void iso_connected(struct bt_iso_chan *chan)
+{
+	LOG_INF("ISO Channel %p connected", chan);
+
+	seq_num = 0U;
+
+	k_sem_give(&sem_big_cmplt);
+}
+
+static void iso_disconnected(struct bt_iso_chan *chan, uint8_t reason)
+{
+	LOG_INF("ISO Channel %p disconnected with reason 0x%02x", chan, reason);
+	k_sem_give(&sem_big_term);
+}
+
+static struct bt_iso_chan_ops iso_ops = {
+	.connected = iso_connected,
+	.disconnected = iso_disconnected,
+};
+
+static struct bt_iso_chan_io_qos iso_tx_qos = {
+	.sdu = sizeof(uint32_t), /* bytes */
+	.rtn = 1,
+	.phy = BT_GAP_LE_PHY_2M,
+};
+
+static struct bt_iso_chan_qos bis_iso_qos = {
+	.tx = &iso_tx_qos,
+};
+
+static struct bt_iso_chan bis_iso_chan[] = {
+	{
+		.ops = &iso_ops,
+		.qos = &bis_iso_qos,
+	},
+	{
+		.ops = &iso_ops,
+		.qos = &bis_iso_qos,
+	},
+};
+
+static struct bt_iso_chan *bis[] = {
+	&bis_iso_chan[0],
+	&bis_iso_chan[1],
+};
+
+static struct bt_iso_big_create_param big_create_param = {
+	.num_bis = 1,
+	.bis_channels = bis,
+	.interval = BIG_SDU_INTERVAL_US, /* in microseconds */
+	.latency = 10, /* in milliseconds */
+	.packing = 0, /* 0 - sequential, 1 - interleaved */
+	.framing = 0, /* 0 - unframed, 1 - framed */
+};
+
+static struct bt_le_ext_adv *adv;
+static struct bt_iso_big *big;
+static uint32_t iso_send_count;
+static uint8_t iso_data[sizeof(iso_send_count)] = { 0 };
+
+struct broadcaster_params broadcast_params = { .sdu_size = 10,
+					       .phy = BT_GAP_LE_PHY_2M,
+					       .rtn = 1,
+					       .num_bis = 1,
+					       .sdu_interval_us = 10000,
+					       .latency_ms = 10,
+					       .packing = 0,
+					       .framing = 0 };
+
+static void broadcaster_t(void *arg1, void *arg2, void *arg3)
+{
+	while (1) {
+		int ret;
+
+		k_sleep(K_USEC(big_create_param.interval));
```

Comment:
Just FYI, if you use the approach of sleeping for the interval here, and the interval is 10000 us, then you won't be sending every interval as the lines below also take time to run. With this approach you are likely only sending every 10100 us (depending on multiple factors) and may miss an SDU interval every 100 SDUs. 

If the test is not sensitive to missed SDUs, this is not a problem, but something to consider. 

The "better" approach is of course to always enqueue 2 SDUs and use the `sent` callback to trigger another TX

Reply:
That is much better. Will be changed.